docs(fx112-css-ray): Create new page for ray() function

dipikabh opened the pull request:
The syntax for the `ray()` function (value in [offset-path](https://developer.mozilla.org/en-US/docs/Web/CSS/offset-path)) has been updated in Fx112.

- `<ray-size>` is now optional.
- The default value of `<ray-size>` is `closest-side`.

### Changes in this PR

- Updates the `ray()` function description in the existing [offset-path](https://developer.mozilla.org/en-US/docs/Web/CSS/offset-path#browser_compatibility) page, adds links to some shape pages

- Adds a new page for the `ray()` function. The webref is not updated with the new syntax so I have not included a "Formal syntax" section. I've added the syntax to the "Syntax" section for now.

Spec: https://drafts.fxtf.org/motion/#ray-function
Firefox bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1820071

### Support in browsers
- Firefox: `ray()` is available behind the flag `layout.css.motion-path-ray.enabled `.
- Chrome: some `offset-path: ray()` functionality is available with "Experimental Web Platform features" enabled but the syntax implementation is not in sync with the latest spec. The BCD table on [offset-path](https://developer.mozilla.org/en-US/docs/Web/CSS/offset-path#browser_compatibility)[ page](https://developer.mozilla.org/en-US/docs/Web/CSS/offset-path#browser_compatibility) contains a row on `ray()` that indicates complete support, which is contrary to what I saw in testing.

### Codepens
Codepens for the examples used in the doc can be found here (tested in Firefox behind the flag):
Example 1: https://codepen.io/dipikabh/pen/yLReQaZ
Example 2: https://codepen.io/dipikabh/pen/PoyNQdz

### Caveat
The functionality is still undergoing discussions:
  - The `contain` keyword behavior is due to be fixed in Firefox (https://bugzilla.mozilla.org/show_bug.cgi?id=1821450) as a result of https://github.com/w3c/fxtf-drafts/issues/363.
  - More change in behavior is expected to come via https://github.com/w3c/fxtf-drafts/issues/504.

### To Do

- [x] Open BCD PR - https://github.com/mdn/browser-compat-data/pull/19421
- [x] Update mdn/data - https://github.com/mdn/data/pull/661
- [x] Update Fx112 release note - https://github.com/mdn/content/pull/26275

### Related issues
Doc issue tracker: https://github.com/mdn/content/issues/25358

dipikabh committed:
copy path to create ray file

dipikabh committed:
add links to basic shape functions

dipikabh committed:
update content

dipikabh committed:
add more info to desc

dipikabh committed:
refine ray definition

dipikabh committed:
add examples

dipikabh committed:
Merge branch 'main' into fx112-ray-function

estelle commented on files/en-us/web/css/offset-path/index.md:
```diff
@@ -17,19 +17,19 @@ The **`offset-path`** [CSS](/en-US/docs/Web/CSS) property specifies a motion pat
 offset-path: none;
 
-/* Function values */
+/* Line segment */
```
if `ray(45deg)` is supposed to work, if values can default, we should include it as an example. (Defaulting does not work in Safari, but i saw this in your other PR).

estelle commented on files/en-us/web/css/offset-path/index.md:
```diff
@@ -43,17 +43,12 @@ offset-path: unset;
-- `ray()`
-  - : Taking up to three values, defines a path that is a line segment starting from the position of the box and proceeds in the direction defined by the specified angle similar to the CSS gradient angle where `0deg` is up, with positive angles increasing in the clockwise direction, with the size value being similar to the CSS radial gradient size values from `closest-side` to `farthest-corner`,  ...
+- {{cssxref("ray","ray()")}}
+  - : Defines a path that is a line segment from the starting position of an element and proceeding in the direction defined by the specified angle. This value accepts up to three parameters – an angle similar to the CSS gradient angle, a size value similar to the CSS radial gradient size values, and the keyword `contain`. The angle `0deg` starts on the y-axis pointing up, and positive angles inc ...
```
```suggestion
  - : Defines a path that is a line segment from the starting position of an element and proceeding in the direction defined by the specified angle. This value accepts up to three parameters – an {{CSSxRef("angle")}} similar to the CSS [linear gradient](/en-US/docs/Web/CSS/gradient/linear-gradient) angle, a size value similar to the CSS [radial gradient](/en-US/docs/Web/CSS/gradient/radial-gradient) size values, and the keyword `contain`. The angle `0deg` starts on the y-axis, pointing up, with positive angles increasing in the clockwise direction. The possible values for the optional size include `closest-side`, `closest-corner`, `farthest-side`, `farthest-corner`, and `sides`, with the size value defaulting to `closest-side` when omitted. The optional `contain` keyword reduces the length of the offset path so that the element stays within its containing block.
```

dipikabh replied:
Thanks!

In the "Note", I am leaving out the second part because in the previous implementation, there was no default for the size parameter. Also I don't think I understand the bit "but has better support."

estelle commented on files/en-us/web/css/ray/index.md:
```diff
@@ -0,0 +1,280 @@
+
+{{CSSRef}}
+
+The **`ray()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Functions) defines a line segment that begins from an [offset starting position](https://drafts.fxtf.org/motion/#offset-starting-position) and extends in the direction of the specified angle. The line segment is referred to as "ray". The length of a ray can be constrained by specifying a size.
```
```suggestion
The **`ray()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Functions) defines a line segment that begins from a starting position set by {{CSSXref("offset-position")}} and extending in the direction of the specified angle. The line segment is referred to as "ray". The length of a ray can be constrained by specifying a size and the keyword `contain`.
```

dipikabh replied:
I intentionally left out linking to `offset-position` page on MDN because the property is [currently not supported](https://developer.mozilla.org/en-US/docs/Web/CSS/offset-position#browser_compatibility) on any browser (and therefore not used in any example) but the concept is supported in calculations. I took cue from the [offset-position](https://developer.mozilla.org/en-US/docs/Web/CSS/offset-position) description that contains a link to spec's "initial position" definition to add a similar link here to a definition in the spec for the offset starting position.

estelle commented on files/en-us/web/css/ray/index.md:
```diff
@@ -0,0 +1,280 @@
+    `sides`: Specifies the length of the line segment as the distance between the ray's starting point and the point where the line segment intersects the containing block's boundary. If the starting point is on or outside the containing block's boundary, the length of the line segment is zero.
+
+- `contain`
+  - : Reduces the length of the line segment so that the element stays within the containing block even at `offset-distance: 100%`. Specifically, the segment's length is reduced by half the width or half the height of the element's border box, whichever is greater.
```
```suggestion
  - : Reduces the length of the line segment so that the element stays within the containing block even at `offset-distance: 100%`. Specifically, the segment's length is reduced to a maximum of half the width or half the height of the element's border box, whichever is greater.
```

estelle commented on files/en-us/web/css/ray/index.md:
```diff
@@ -0,0 +1,280 @@
+
+.box2 {
+  offset-path: ray(0deg);
+  offset-anchor: 0px 0px;
```
Should we do `25%` on some? `px` is not needed for lengths of `0`, and many best practices suggest omitting them. This comment is for all the example boxes.

estelle commented on files/en-us/web/css/ray/index.md:
````diff
@@ -0,0 +1,280 @@
+```html hidden
+<div class="container">
+  <div class="example">
+    <div class="box">&nbsp;</div>
````
instead of &nbsp;, write "box1" or "1" to make the examples easier to grok.

estelle commented on files/en-us/web/css/ray/index.md:
````diff
@@ -0,0 +1,280 @@
+</div>
+```
+
+The default anchor point is at the center of an element, which can be modified using the [`offset-anchor`](/en-US/docs/Web/CSS/offset-distance) property. The default offset starting position of a path at is the top-left corner of the element box.
````
```suggestion
Similar to {{cssxref("transform-origin")}}, the default anchor point is at the center of an element. This anchor point can be modified using the [`offset-anchor`](/en-US/docs/Web/CSS/offset-distance) property. The default offset starting position of a path at is the top-left corner of the element box.
```

estelle reviewed (commented):
a few comments. looking good. still working on the review.

estelle commented on files/en-us/web/css/ray/index.md:
```diff
@@ -0,0 +1,280 @@
+  </div>
+</div>
+
+<pre>offset-path: ray(120deg sides contain);<br>offset-rotate: 0deg;</pre>
```
```suggestion
<pre>offset-path: ray(120deg sides contain);
offset-rotate: 0deg;</pre>
```

estelle commented on files/en-us/web/css/ray/index.md:
```diff
@@ -0,0 +1,280 @@
+
+{{EmbedLiveSample('Animating an element along the ray', '100%', 750)}}
+
+The first shape is shown as a reference for its position and orientation. A ray motion path is applied on the other shapes. In the first two `offset-path` samples, notice the orientation of the shape with and without `offset-rotate`.
```
```suggestion
The first shape is shown as a reference for its position and orientation. A ray motion path is applied on the other shapes. In the first two `offset-path` samples, notice the orientation of the shape with and without `offset-rotate`. The last two show the impact of size: `closest-corner` creates the a `0px` long offset path as the shape is already in the corner. The `farthest-corner` value creates the longest offset-path, going from the top left corner of the container to the bottom right corner in this case.
```

estelle replied:
My version: I updated the example a bit: https://codepen.io/estelle/pen/XWxjzoQ

estelle reviewed (commented):
made some suggestions. Also forked your codepens. The first example doesn't work for me without size.

dipikabh commented:
Thanks for the review @estelle and for your help with fixing the code.

1. I have fixed your comments except I wasn't able to move part of CSS in second example to "css hidden" (Firefox 112 with the flag `layout.css.motion-path-ray.enabled`):
- In the first example, I put part of the CSS in "css hidden" like you suggested and the example works fine.
- In the second example, when I put part of CSS code in "css hidden", the animation stopped working (while I could see the animation working in your codepen in the same browser, different tab). The animation started working after I removed "css hidden".

2. I added `closest-side` to all cases in the first example because I could validate the "Note" text you've added. Just to make a record of the observations here:
  - specifying only the angle value is working in Firefox 112 behind the flag
  - specifying only the angle value is working in Chrome Canary with `#enable-experimental-web-platform-features`
  - not working in Chrome 112 with or without `#enable-experimental-web-platform-features` (in exp mode, first example works with explicit mention of `closest-side`, second example still not working)
  - second example is working in Safari TP, indicating that `sides` might be better supported than `closest-side`? `offset-anchor` does not work

dipikabh committed:
fix review comments

dipikabh committed:
Merge branch 'main' into fx112-ray-function

estelle reviewed (approved):
Thanks! 🎉
